circe-display-images.el

```diff
@@ -0,0 +1,197 @@
+  "A hash map used to manage display transitions.
+
+The keys are urls, and the values are a plist with an `:image-property', and a
+`:display-image-p'. `:image-property' is the display property of the image, and
+`:display-image-p' is a flag telling us whether the image is currently visible
+or not. This map serves to keep track of display transitions, and as a mapping
+between the URL and its downloaded image.
+
+Unfortunately we can't map from URL to the image position in the buffer
+because 1) the lui library can move text around when executing the
+`lui-post-output-hooks' and 2) as we toggle images, that also changes other
+images' position in the buffer.")
+
+(defun circe-display-images-toggle-image-at-point ()
+  "Toggle the image corresponding to the url at point.
+
+This function iterates through all display properties in the buffer. We look
+for a match with the display property we got from our property map, with the
+url-at-point as the key. When we find a match, we either remove or add back
+the image. See `circe-display-images-text-property-map' for more details."
+  ;; Giant thank you to Malabarba who's S-O answer I slightly modified:
+  ;; https://emacs.stackexchange.com/a/566
+  (interactive)
+  (let*
+      ((inhibit-read-only t)
+       (url (url-get-url-at-point))
+       (image-data(gethash url circe-display-images-text-property-map))
+       (display-image-p (plist-get image-data :display-image-p))
+       (image-property-of-url (plist-get image-data :image-property))
+       (from (if display-image-p 'display 'display-backup))
+       (to (if display-image-p 'display-backup 'display))
+       (current-pos (point-min))
+       left current-image-property)
+    (while (and current-pos (/= current-pos (point-max)))
+      ;; Find the next image property in the buffer.
+      (if (get-text-property current-pos from)
+          (setq left current-pos)
+        (setq left (next-single-property-change current-pos from)))
+      (if (or (null left) (= left (point-max)))
+          (setq current-pos nil)
+        (setq current-image-property (get-text-property left from))
+        (setq current-pos (or (next-single-property-change left from)
+                              (point-max)))
+        ;; Swap the images if our current image matches the image from the URL.
+        (when (equal image-property-of-url current-image-property)
+          (add-text-properties
+           left current-pos (list from nil to current-image-property)))))
+    ;; Make sure to invert the :display-image-p flag after processing all
+    ;; images.
+    (puthash url `(:image-property ,image-property-of-url
+                   :display-image-p ,(not display-image-p))
+             circe-display-images-text-property-map)))
+
+(defun circe-display-images-insert-image-from-url (url)
+  "Attempt to download the image from URL, and insert it."
+  (let ((buffer (url-retrieve-synchronously url)))
```

Comment:
This should presumably be rewritten to use url-retrieve, blocking is particularly bad in an IRC client.

Reply:
Thanks. I tried implementing this originally but I was running into some strange performance issues when using `url-retrieve`. In the example below, the asynchornous command seems to make emacs less responsive than the synchronous one.

You can try this by evaluating each expression yourself. Does this behave the same way on your machine? Is this expected behavior? Maybe I'm misunderstanding how url-retrieve is supposed to work?
```lisp
(let* (( url (url-retrieve-synchronously "https://i.imgur.com/xmm4TSh.jpg"))
       (data (with-current-buffer buffer (goto-char (point-min)) (search-forward "\n\n")
                                  (buffer-substring (point) (point-max)))))
  (print "done"))

(url-retrieve "https://i.imgur.com/xmm4TSh.jpg"
              (lambda (s)
                (let ((data (with-current-buffer buffer (goto-char (point-min)) (search-forward "\n\n")
                                                 (buffer-substring (point) (point-max))))) (print "done"))))
                                                  
```

Reply:
@wasamasa any advice regarding the asynchronous url-retrieve function? I agree that we should absolutely not block and url-retreive seems like the right solution, I'm just confused why in the above code the asynchronous call is slower. I'm assuming I'm misunderstanding the API somehow.

Reply:
Your benchmark code is faulty because `benchmark` is a function, so the form passed to it must be quoted. That asides, benchmarking a network operation is rather pointless, all it ends up doing is for the first one to measure how long I need to download 100 images (more than a minute!) and for the second one how long it takes to make such a connection (not quite a minute) and then spams my minibuffer with messages.

Reply:
Sure, but my original point was even evaluating these without the benchmark, i.e:

``` lisp
(require 'url)

(let* ((buffer (url-retrieve-synchronously "https://i.imgur.com/xmm4TSh.jpg"))
                    (data (with-current-buffer buffer (goto-char (point-min)) (search-forward "\n\n")
                                  (buffer-substring (point) (point-max)))))
               (print "hello"))

(url-retrieve
                "https://i.imgur.com/xmm4TSh.jpg"
                (lambda (s)
                  (let ((data (with-current-buffer (current-buffer) (goto-char (point-min)) (search-forward "\n\n")))))
                  (buffer-substring (point) (point-max)) (print "hello")))
```
I'm still not seeing what I expected. What I expect is to immediately 
get control back of my cursor when I execute the asynchronous version,
but instead it ( seems to ) takes as much time as the synchronous version.

Isn't the point of using `url-retrieve` over `url-retrieve-synchronously` is
so that we don't block if the network connection is slow? Somehow the way
I'm using `url-retrieve` isn't accomplishing this in the above trivial example.

Reply:
I cannot reproduce this.  What I learned from chats about Circe is that a blocking part of creating network processes is the DNS lookup, so creating the connection may still block.  Perhaps @jorgenschaefer can say more about this, assuming it's still a thing (Lars supposedly handed in a patch to fix it).

Reply:
I have no idea if that was fixed. It probably greatly depends on what kind of libraries are available. The standard C ones are blocking, and would require threading I think. "Nothing you can do about it", I'm afraid.

Reply:
I was not able to implement this without blockign using `url-retrieve`. I'll leave it as is, but I'll keep playing around with an async implementation ( and keep an eye out for helpful PRs 😊 )

Reply:
@naclander `request` is non-blocking, including DNS lookups, when set to use `curl`.

Reply:
@naclander I have difficulties understanding this reasoning.  `url-retrieve-synchronously` is by definition worse than `url-retrieve` (go look at its code, I dare you), yet you still use `url-retrieve-synchronously` because `url-retrieve` doesn't give you a better user experience?